### PR TITLE
perf(quota): Skip expensive usage checks when no quota configured

### PR DESCRIPTION
**Pull request by thorntonmc**
## Type of Change
- [ ] New Feature
- [x] Bug Fix
- [ ] Documentation
- [ ] Performance Improvement
- [ ] Test/CI
- [ ] Refactor
- [ ] Other:

## Related Issues
#1748 

## Summary of Changes
Bucket quota checks are performed on every object operation (put, copy, multipart upload), even when no quota is configured for the bucket. This causes unnecessary performance overhead as the system calculates current bucket usage for every operation, regardless of whether quota enforcement is needed.

As this check is also a sync check in the write path, this can **completely block writes** while the quota check runs. This was particularly noticeable when load testing against our RustFS instance.

Solution
This PR introduces a performance-optimized quota checking system that eliminates duplicate operations while maintaining API accuracy:

Key changes:

1. **Option<u64> for current_usage**: Modified `QuotaCheckResult.current_usage` from `u64` to `Option<u64>` to be honest about when usage calculation is skipped for performance
2. **Specialized admin method**: Added `check_quota_with_usage_reporting()` that forces usage calculation when admin APIs need precise data
3. **Fast path optimization**: Skip expensive usage calculation entirely when no quota is configured
4. **API transparency**: The API now clearly indicates when usage calculation was skipped (`None`) vs. when it was calculated (`Some(value)`)

Performance Benefits:
* Zero-quota buckets: Skip expensive usage calculations entirely
* Quota-enabled buckets: No performance impact, same behavior as before
* Error cases: Graceful fallback with warning logs
* Existing quota tests continue to pass
* New method handles edge cases (empty config, parse errors)
* Usage tracking still works for both quota and non-quota buckets

## Checklist
- [ ] I have read and followed the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines
- [ ] Passed `make pre-commit`
- [ ] Added/updated necessary tests
- [ ] Documentation updated (if needed)
- [ ] CI/CD passed (if applicable)

## Impact
- [ ] Breaking change (compatibility)
- [ ] Requires doc/config/deployment update
- [ ] Other impact:

## Additional Notes

---

### CODE REVIEW

**Comment by houseme** [comment]
Thank you for your contribution. It is recommended to determine this inside `quota_checker.check_quota` to minimize modifications.

**Comment by thorntonmc** [comment]
> Thank you for your contribution. It is recommended to determine this inside `quota_checker.check_quota` to minimize modifications.

Thank you for pointing out the performance regression around the extra metadata call and json serialization. To avoid that, while still keeping the performance benefit of not calculating usage in the write path I have kept the logic within `check_quota`, which now has been broken out into `check_quota_with_usage_reporting`. This is to keep compatibility with the admin panel, which relies on having usage calculated for buckets without quotas. 

I have also modified `QuotaCheckResult.current_usage` from `u64` to `Option<u64>` to be honest about when usage calculation is skipped for performance.

File: crates/ecstore/src/bucket/quota/checker.rs [hunk 1]
```diff
@@ -36,14 +36,30 @@ impl QuotaChecker {
...
             None => {
-                let current_usage = self.get_real_time_usage(bucket).await?;
+                let current_usage = if force_usage_calculation {
+                    Some(self.get_real_time_usage(bucket).await?)
+                } else {
+                    None // Skip expensive usage calculation when no quota and not forced for performance
+                };
                 return Ok(QuotaCheckResult {
                     allowed: true,
                     current_usage,
```
[the remark this answers is not in the document]
Comment (thorntonmc) [reply crates/ecstore/src/bucket/quota/checker.rs:65]: @houseme can we seed the cache when enabling the quota instead of here? Having this be in the fast path can degrade write performance in our case, we were unable to write at all for several minutes while multiple instances of `get_real_time_usage` were burning up CPU usage.

[the remark this answers is not in the document]
Comment (houseme) [reply crates/ecstore/src/bucket/quota/checker.rs:65]: Multiple machines and multiple disks can cause data overage situations? It is only suitable for single-machine scenarios, but the indicator for whether quotas are enabled can be handled using caching.

[the remark this answers is not in the document]
Comment (thorntonmc) [reply crates/ecstore/src/bucket/quota/checker.rs:65]: @houseme yes you are correct. I now no longer call `increment_bucket_usage_memory()` for non-quota buckets preventing the cache corruption pointed out by Copilot.

When a bucket then has a quota enabled, the cache will miss on all machines and the quotas will be prperly enforced.

[the remark this answers is not in the document]
Comment (yxrxy) [reply crates/ecstore/src/bucket/quota/checker.rs:65]: The Data Scanner periodically scans the actual usage of all buckets and saves it to the backend (crates/scanner/src/scanner.rs:43-61,164-234). The in-memory cache has
   a 30-second TTL, and when the cache expires, get_bucket_usage_memory automatically loads the latest data from the backend (crates/ecstore/src/data_usage.rs:407-481). Therefore, even
  in the "no quota" scenario where the fast path is taken, the periodic data scanner scanning and TTL automatic expiration mechanism ensure eventual consistency. There is no issue where
  the cache would be permanently lower than the actual value.

  Recommendation: You can ignore this Copilot suggestion. For further memory usage optimization, consider checking whether the bucket has quota enabled in write/delete operations, and
  only call increment_bucket_usage_memory/decrement_bucket_usage_memory when quota is enabled. For buckets without quota, skip the in-memory cache update entirely. This can significantly
   reduce memory usage (by avoiding maintaining in-memory cache for all buckets), while relying on the periodic data scanner scanning and TTL automatic expiration mechanism to ensure
  eventual consistency. When dynamically enabling quota, there will be at most a 30-second delay (TTL) + scanner scan cycle, which is an acceptable trade-off since quota configuration is
   typically not frequently changed.
